core/src/main/scala/kafka/server/ReplicaFetcherThread.scala

```diff
@@ -343,16 +345,21 @@ class ReplicaFetcherThread(name: String,
       .filter { case (_, state) => state.isTruncatingLog }
       .map { case (tp, _) => tp -> epochCacheOpt(tp) }.toMap
 
-    val (partitionsWithEpoch, partitionsWithoutEpoch) = partitionEpochOpts.partition { case (_, epochCacheOpt) => epochCacheOpt.nonEmpty }
+    val (partitionsWithCache, partitionsWithoutCache) = partitionEpochOpts.partition { case (_, epochCacheOpt) => epochCacheOpt.nonEmpty }
+    val (partitionsWithEpoch, partitionsWithoutEpoch) = partitionsWithCache
+      .map { case (tp, epochCacheOpt) => tp -> epochCacheOpt.get.latestEpoch() }
+      .partition { case (_, latestEpoch ) => latestEpoch != UNDEFINED_EPOCH }
+
+    // partitions for which it makes sense to request epoch
+    val partitionsWithError = (partitionsWithoutCache.keys ++ partitionsWithoutEpoch.keys).filter { tp => topicSupportsEpochRequest(tp) }.toSet
+    debug(s"Build leaderEpoch request: Partitions with epoch: $partitionsWithEpoch; Partitions without epoch: $partitionsWithError")
 
-    debug(s"Build leaderEpoch request $partitionsWithEpoch")
-    val result = partitionsWithEpoch.map { case (tp, epochCacheOpt) => tp -> epochCacheOpt.get.latestEpoch() }
-    ResultWithPartitions(result, partitionsWithoutEpoch.keys.toSet)
+    ResultWithPartitions(partitionsWithEpoch, partitionsWithError)
```

Comment:
I think the desired behavior is for partitions which do not support the latest message format to fall back to truncation using the high watermark. However, `partitionsWithError` ultimately get passed to `handlePartitionsWithErrors`, which then calls `delayPartitions`. That seems not quite what we want?